test/BulletDynamics/test_deterministic_save_and_restore.cpp

```diff
@@ -0,0 +1,221 @@
+
+	btCollisionDispatcher * dispatcher = new btCollisionDispatcher(collisionConfiguration);
+
+	btDbvtBroadphase * broadphase = new btDbvtBroadphase();
+
+	btMultiBodyConstraintSolver* solver = new btMultiBodyConstraintSolver();
+
+	btMultiBodyDynamicsWorld * world = new btMultiBodyDynamicsWorld(dispatcher, broadphase, solver, collisionConfiguration);
+
+	world->setGravity(btVector3(0, -10, 0));
+
+	return world;
+
+}
+
+GTEST_TEST(BulletDynamics, DeterministicSaveRestore)
+{
+
+	static char filename[] = "test_serialize.mixail";
+	btRigidBody* box_before[3];
+	btCollisionObject* box_after[3];
+	btMultiBodyDynamicsWorld *initial_world = create_btMultiBodyDynamicsWorld();
+
+	// create ground plane and box shapes
+	// don't worry about gui things
+
+	btBoxShape* groundShape = new btBoxShape(btVector3(btScalar(50.), btScalar(50.), btScalar(50.)));
+	{
+		btTransform groundTransform;
+		groundTransform.setIdentity();
+		groundTransform.setOrigin(btVector3(0, -50, 0));
+		btScalar mass(0.);
+		bool isDynamic = (mass != 0.f);
+		btVector3 localInertia(0, 0, 0);
+		if (isDynamic)
+			groundShape->calculateLocalInertia(mass, localInertia);
+
+		#ifdef USE_MOTIONSTATE
+			btDefaultMotionState* myMotionState = new btDefaultMotionState(groundTransform);
+
+			btRigidBody::btRigidBodyConstructionInfo cInfo(mass, myMotionState, groundShape, localInertia);
+
+			btRigidBody* body = new btRigidBody(cInfo);
+		#else
+			btRigidBody* body = new btRigidBody(mass, 0, groundShape, localInertia);
+			body->setWorldTransform(groundTransform);
+		#endif
+
+		body->setUserIndex(-1);
+		initial_world->addRigidBody(body);
+	}
+	{
+
+		btBoxShape* colShape = new btBoxShape(btVector3(.1, .1, .1));
+
+		btTransform startTransform;
+		startTransform.setIdentity();
+
+		btScalar mass(1.f);
+
+		bool isDynamic = (mass != 0.f);
+
+		btVector3 localInertia(0, 0, 0);
+		if (isDynamic)
+			colShape->calculateLocalInertia(mass, localInertia);
+		for (int i = 0; i < 3; i++)
+		{
+			int a[4];
+			if (i == 0)
+			{
+				a[0] = 2;
+				a[1] = 0;
+				a[2] = -1;
+				a[3] = 1;
+			}
+			else if (i ==1)
+			{
+				a[0] = 2;
+				a[1] = 2;
+				a[2] = -1;
+				a[3] = -1;
+			}
+			else
+			{
+				a[0] = 0;
+				a[1] = 1;
+				a[2] = 1;
+				a[3] = 0;
+			}
+			startTransform.setOrigin(btVector3(
+				btScalar(a[0]),
+				btScalar(10),
+				btScalar(a[1])));
+			#ifdef USE_MOTIONSTATE
+				btDefaultMotionState* myMotionState = new btDefaultMotionState(startTransform);
+
+				btRigidBody::btRigidBodyConstructionInfo cInfo(mass, myMotionState, colShape, localInertia);
+
+				box_before[i] = new btRigidBody(cInfo);
+			#else
+				box_before[i] = new btRigidBody(mass, 0, colShape, localInertia);
+				box_before[i]->setWorldTransform(startTransform);
+			#endif
+			box_before[i]->setUserIndex(-1);
+			initial_world->addRigidBody(box_before[i]);
+			box_before[i]->setLinearVelocity(btVector3(a[2], 0, a[3]));
+		}
```

Comment:
If you're adding 3 boxes, I would just add 3 boxes in a row, or better yet write a function to create a box at a given location and with a particular velocity. This for loop with the weird 'a' variable is rather cumbersome to read.